### DISCUSSION TITLE
expose publicClient on contract

### DISCUSSION THREAD

**Discussion by sakulstra** [wevm/viem#722] [Idea / Feature Request]
Hello,

I'm using viem in a multichain project where one contract is deployed on multiple networks.
Viem is then used to fetch data, cache events etc.

To create a contract instance I would then do:
```
const arbitrumContract = getContract({
  address: ARBITRUM_ADDRESS,
  abi: ABI,
  publicClient: arbitrumClient,
});
```

The issue i'm facing though is that when i e.g. create a generic method i might want to also cache the latest block etc, but `arbitrumContract` doesn't expose a way to access the chainId or the client itself.

Ofc, I can work around this, but wondering if it would be rational to just expose the `publicClient`  similar to what i did here: https://github.com/wagmi-dev/viem/pull/619

**Comment by sakulstra** [comment]
Actually related to that, in some occasions it's handy to know the rpc url of the client as e.g. tenderly has other methods available then anvil. So it might make sense in my application logic to check which kind of rpc is connected - i think it's not currently possible to read the rpc from the `transport`. WOuld be handy if that would somehow be exposed as well.

**Comment by Sam-Scolari** [comment]
I second this!

I would like to be able to do things like this

```typescript
// Waiting for a transaction
const hash = await contract.write.castVote([proposalId]);
const transaction = await contract.publicClient.waitForTransactionReceipt({ hash })

// Getting the chain the contract exists on
import { sepolia } from "viem/chains";
const chain = contract.publicClient.chain;
chain === sepolia // true
```

Would also be nice to expose the wallet client.
